From 0c1db29cb40f795e637acce3062684e4e3d9e89b Mon Sep 17 00:00:00 2001
From: Ivan Vecera <ivecera@redhat.com>
Date: Wed, 19 Nov 2014 16:14:11 -0500
Subject: [ethernet] bna: fix skb->truesize underestimation

Message-id: <1416413651-12167-4-git-send-email-ivecera@redhat.com>
Patchwork-id: 100348
O-Subject: [RHEL7.1 PATCH 3/3] bna: fix skb->truesize underestimation
Bugzilla: 1165759
RH-Acked-by: Dean Nelson <dnelson@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

BZ: #1165759

Upstream commit(s):
commit f2d9da1a8375cbe53df5b415d059429013a3a79f
Author: Eric Dumazet <edumazet@google.com>
Date:   Fri Oct 17 12:45:55 2014 -0700

    bna: fix skb->truesize underestimation

    skb->truesize is not meant to be tracking amount of used bytes
    in an skb, but amount of reserved/consumed bytes in memory.

    For instance, if we use a single byte in last page fragment,
    we have to account the full size of the fragment.

    skb->truesize can be very different from skb->len, that has
    a very specific safety purpose.

    Signed-off-by: Eric Dumazet <edumazet@google.com>
    Cc: Rasesh Mody <rasesh.mody@qlogic.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Ivan Vecera <ivecera@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/brocade/bna/bnad.c b/drivers/net/ethernet/brocade/bna/bnad.c
index d3307ea..be74a43 100644
--- a/drivers/net/ethernet/brocade/bna/bnad.c
+++ b/drivers/net/ethernet/brocade/bna/bnad.c
@@ -552,6 +552,7 @@ bnad_cq_setup_skb_frags(struct bna_rcb *rcb, struct sk_buff *skb,
 
   len = (vec == nvecs) ?
    last_fraglen : unmap->vector.len;
+  skb->truesize += unmap->vector.len;
   totlen += len;
 
   skb_fill_page_desc(skb, skb_shinfo(skb)->nr_frags,
@@ -563,7 +564,6 @@ bnad_cq_setup_skb_frags(struct bna_rcb *rcb, struct sk_buff *skb,
 
  skb->len += totlen;
  skb->data_len += totlen;
- skb->truesize += totlen;
 }
 
 static inline void
-- 
1.7.1